[AERIE-1726]  Activity Typescript Types Generation

dyst5422 opened the pull request:
* **Tickets addressed:** AERIE-1726
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description

This PR will generate the activity_type's typescript Library used to author the command expansions.  Here is the workflow:

1. Use a Hasura action and provide the `missionModelID` and `Activity_Type's` name. The action will provide this information to the commanding server
2. The commanding server will retrieve the parameters from the activity_type
3. Once the parameters are queried, a typescript library is generated
4. The library is base64 encoded and sent back as the response to the Hasura action

## Verification
I test locally without any issues using `ParmeterTest` activity_type, which has very complicated parameters objects. 

1. `docker system prune --all --volumes` nuke everything 
2. `./gradlew clean; ./gradlew assemble; docker compose -f docker-compose.yml build; docker compose -f docker-compose.yml up` 
3. Add a mission model jar
4. Run the action which has the activity_type name and mission model 

```
query MyQuery {
  getActivityTypeScript(activityTypeName: "BakeBananaBread", missionModelId: "1") {
    typescript
  }
}
```
5. Verify that you receive `base64` string back from the action.
6. Take the string and convert it with this URL: https://www.base64decode.org/


## Documentation
I will have to add these steps to the User Guide that we will be giving the users

## Future work
Creating Hasura action for the other endpoints in the commanding server and hooking up the rest of the commanding server.

dyst5422 commented:
@goetzrrGit let me know when you think you have this pretty well put together, and I'm going to go through and rebase to cleanup and ensure it builds on our other PRs for merging later.

goetzrrGit commented:
@dyst5422 This is pretty much done. You can probably `rebase` and `drop `the commits before https://github.com/NASA-AMMOS/aerie/commit/17125226521bc925bad83305bb7fefffdcd706fc. That should clean up this PR and only reflect the new changes. Once the Upload PR is merged It should be rebased into this PR with minor issues.

I remove the other endpoints that are not part of the PR. Also, I didn't include dev environment stuff as I think that can be part of another localized PR. It will just help so the PR isn't so big

dyst5422 commented:
@goetzrrGit Yeah, the intent with these PR's that I put together was to put together a merging plan ahead of time and still providing the niceties of some of the other cleanup and prep work.

So the merge plan was #78 -> [#80, #81, #82, #82] -> The execution PR which builds on all of them.

So I'm going to rebase this on top of #78 and probably cherry pick out lines and pull commits together to tell a cohesive story without all of the development time alterations (unused package removal for example as we switched the api from a form-data to json)

goetzrrGit commented:
@dyst5422 Sounds like a good idea. Let me know if you need any help?

goetzrrGit commented on command-expansion-server/src/getActivityTypescript.ts:
```diff
@@ -0,0 +1,131 @@
+  Variant = 'variant',
+}
+
+type Schema =
+  | IntSchema
+  | RealSchema
+  | DurationSchema
+  | BooleanSchema
+  | StringSchema
+  | SeriesSchema
+  | StructSchema
+  | VariantSchema
+  ;
+
+interface BaseSchema<T extends SchemaTypes | unknown> {
+  type: T,
+}
+
+type StringSchema = BaseSchema<SchemaTypes.String>;
+type BooleanSchema = BaseSchema<SchemaTypes.Boolean>;
+type IntSchema = BaseSchema<SchemaTypes.Int>;
+type RealSchema = BaseSchema<SchemaTypes.Real>;
+type DurationSchema = BaseSchema<SchemaTypes.Duration>;
+
+interface SeriesSchema extends BaseSchema<SchemaTypes.Series> {
+  items: Schema,
+}
+
+interface StructSchema extends BaseSchema<SchemaTypes.Struct> {
+  items: {
+    [key: string]: Schema,
+  }
+}
+
+interface VariantSchema extends BaseSchema<SchemaTypes.Variant> {
+  variants: {
+    key: string,
+    label: string,
+  }[]
+}
```
You are at Typescript God!!!!

Super minor, can you order this to be top-down declaration? It helps with readability.

```
enum SchemaTypes {
  Int = "int",
  Real = "real",
  Duration = "duration",
  Boolean = "boolean",
  String = "string",
  Series = "series",
  Struct = "struct",
  Variant = "variant",
}

interface BaseSchema<T extends SchemaTypes | unknown> {
  type: T;
}
interface SeriesSchema extends BaseSchema<SchemaTypes.Series> {
  items: Schema;
}

interface StructSchema extends BaseSchema<SchemaTypes.Struct> {
  items: {
    [key: string]: Schema;
  };
}

interface VariantSchema extends BaseSchema<SchemaTypes.Variant> {
  variants: {
    key: string;
    label: string;
  }[];
}

type StringSchema = BaseSchema<SchemaTypes.String>;
type BooleanSchema = BaseSchema<SchemaTypes.Boolean>;
type IntSchema = BaseSchema<SchemaTypes.Int>;
type RealSchema = BaseSchema<SchemaTypes.Real>;
type DurationSchema = BaseSchema<SchemaTypes.Duration>;


type Schema =
  | IntSchema
  | RealSchema
  | DurationSchema
  | BooleanSchema
  | StringSchema
  | SeriesSchema
  | StructSchema
  | VariantSchema;
```

dyst5422 replied:
I think because TS doesn't care much about WHERE the declarations are, many libraries do it this way so that the root or most important types are higher up (seen sooner when looking through the file) and then the needed children come after them.

I'm good either way though, feel free to let me know which you prefer given the above.

dyst5422 replied:
See https://github.com/graphql/graphql-js/blob/main/src/language/ast.ts#L139 as a good example of general typing patterns

goetzrrGit reviewed (approved):
You made this way better.

goetzrrGit committed:
Added the activity typescript generation class

This will fetch an activities parameters from hasura and convert them into typscript library.

The package.json has been updated to include a graphql library

goetzrrGit committed:
Setup the Hasura environment variable so the commanding server know which url to make the graphQL call.

goetzrrGit committed:
Update the app.post("/activity-typescript") endpoint.

input:

{
  "input": {
    "missionModelId": Int,
    "activityTypeName": String
  }
}

and returns the activity typescript base64 encoded

{
  "typescript": Base64EncodedString
}

goetzrrGit committed:
Add new Hasura action for getActivityTypeScript

This action will allow users to generate an activity's typescript library file.

pcrosemurgy commented on command-expansion-server/src/getActivityTypescript.ts:
```diff
@@ -0,0 +1,131 @@
+  console.log(`query parameters from ${activityTypeName} with missionModelId: ${missionModelId}`);
+  const response = await graphqlClient.request<{
+    activity_type: GraphQLActivity[]
+  }>(gql`
```
Maybe this is a little pedantic but the mixing of GraphQL querying and code-gen. logic within the same "translation unit" doesn't feel quite right. I can understand how decoupling this querying from generation may feel a bit overkill right now but in my experience having disciplined decoupling early on can save trouble later.

pcrosemurgy replied:
One benefit to decoupling would be testability. Currently, in order to test this code gen. you'd need to have a `GraphQLClient`. Instead, if there was an abstract interface that the code generation could take as input then you could write a series of tests fairly easily.

dyst5422 replied:
@pcrosemurgy 

I think this is the use case for having **generateTypescriptForGraphQLActivitySchema** below. The exported `getActivityTypescript` is an abstraction over both the query and the generation. Until we have a use case for performing this query without generating types, I don't feel the additional complication of extracting it is worthwhile (and incurs runtime costs). I think we can definitely export the type generation function for testing though

pcrosemurgy replied:
That's fair enough, all boils down to personal preference about modularity.

>and incurs runtime costs

I'd imagine a call out to another class/function should be very cheap (<10 ms.?). Either way, all good :)

pcrosemurgy commented on deployment/hasura/metadata/actions.graphql:
```diff
@@ -25,6 +25,17 @@ type Query {
+  ): TypeScriptResponse
+}
+
+type TypeScriptResponse {
```
Another slightly pedantic change: would you mind moving `TypeScriptResponse ` to further down in this file?

Looks like this file has this rough ordering:
- Mutations
- Queries
- Types needed for mutations & queries (`*Response` types)
- Scalars

dyst5422 replied:
I'm not disagreeing with you. I think there is an argument to be made for an alternative "grouping" of each Query/Mutation/Subscription operation with its own relevant types and collapsing shared types elsewhere. Not sure which is more productive, but I know I like to see everything together so its easier to grok

goetzrrGit replied:
To be honest I don't have a preference. The other PR has the rough ordering just to be consistent:

[deployment/hasura/metadata/actions.graphql
](https://github.com/NASA-AMMOS/aerie/pull/80/files#diff-1495af69e68d7d8f4bdaf2b9437fb413026dbee6270f8a7676cc95e102543f50R115)

pcrosemurgy replied:
No strong preference here besides consistency, so feel free to reorder the rest of the actions if you prefer the grouping order.

dyst5422 replied:
Not worth reordering the whole thing. Moved it to follow the other PR

pcrosemurgy commented on command-expansion-server/package.json:
```diff
@@ -21,6 +21,7 @@
     "body-parser": "^1.19.2",
     "express": "^4.17.2",
     "graphql": "^16.3.0",
+    "graphql-request": "^4.0.0",
```
Is the `graphql` dependency being used? If not then maybe it can be fully replaced with the `graphql-request` dep.?

goetzrrGit replied:
yeah, `graphql` is used to build schemas which we are not doing in this part of the server. @dyst5422 where you using this library somewhere else?

dyst5422 replied:
Removed

mattdailis commented on command-expansion-server/src/getActivityTypescript.ts:
```diff
@@ -0,0 +1,131 @@
+  console.log(`query parameters from ${activityTypeName} with missionModelId: ${missionModelId}`);
+  const response = await graphqlClient.request<{
+    activity_type: GraphQLActivity[]
+  }>(gql`
+    query GetParameters {
+      activity_type(where: {
+        model_id: { _eq: ${missionModelId} },
+        name: { _eq: ${activityTypeName} },
+      }) {
+        name
+        parameters
+      }
+    }
+  `);
```
🔻 Would it be possible to use GraphQL variables instead of string interpolation here? It would be a very minimal change, and it would prevent GraphQL injection issues via the activityTypeName:
```suggestion
  }>(gql`
    query GetParameters($missionModelId: Int!, $activityTypeName: String!) {
      activity_type(where: {
        model_id: { _eq: $missionModelId },
        name: { _eq: $activityTypeName },
      }) {
        name
        parameters
      }
    }
  `, {missionModelId, activityTypeName});
```

goetzrrGit replied:
Thanks for the suggestion. I made the change and verified it works as expected

mattdailis commented on command-expansion-server/src/getActivityTypescript.ts:
```diff
@@ -0,0 +1,131 @@
+        parameters
+      }
+    }
+  `);
```
Would it be possible to use GraphQL variables instead of string interpolation here? It would be a very minimal change, and it would prevent GraphQL injection issues via the activityTypeName:
```suggestion
  }>(gql`
    query GetParameters($missionModelId: Int!, $activityTypeName: String!) {
      activity_type(where: {
        model_id: { _eq: $missionModelId },
        name: { _eq: $activityTypeName },
      }) {
        name
        parameters
      }
    }
  `, {missionModelId, activityTypeName});
```

mattdailis commented on command-expansion-server/src/getActivityTypescript.ts:
```diff
@@ -0,0 +1,131 @@
+          .join('\n')
+      );
+    case SchemaTypes.Variant:
+      return `${schema.variants.map(variant => `'${variant.label}'`).join(" | ")}`;
```
🔶 It might be a good idea to wrap the enum in parenthesis - the case I'm thinking of is if there's a list of enums, you don't want to end up with:
```
'A' | 'B' | 'C'[]
```

goetzrrGit replied:
good eyes Matt. I tweaked it to include '()' and tested with the `ParameterTest` activity.

mattdailis commented on command-expansion-server/src/getActivityTypescript.ts:
```diff
@@ -0,0 +1,131 @@
+          .join('\n')
+      );
+    case SchemaTypes.Variant:
+      return `${schema.variants.map(variant => `'${variant.label}'`).join(" | ")}`;
```
🔶 It might be a good idea to wrap the enum in parenthesis - the case I'm thinking of is if there's a list of enums, you don't want to end up with:
```
"A" | "B" | "C"[]
```

jdylanstewart committed:
[DEPS] Remove graphql dependency

jdylanstewart committed:
[REFACTOR] Reorganize graphql actions
